# Проект по SQL

# Описание проекта:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных:

- **Таблица books**
- Содержит данные о книгах:
   - book_id — идентификатор книги;
   - author_id — идентификатор автора;
   - title — название книги;
   - num_pages — количество страниц;
   - publication_date — дата публикации книги;
   - publisher_id — идентификатор издателя.
   
- **Таблица authors**
- Содержит данные об авторах:
   - author_id — идентификатор автора;
   - author — имя автора.
   
- **Таблица publishers**
- Содержит данные об издательствах:
   - publisher_id — идентификатор издательства;
   - publisher — название издательства;
   
- **Таблица ratings**
- Содержит данные о пользовательских оценках книг:
   - rating_id — идентификатор оценки;
   - book_id — идентификатор книги;
   - username — имя пользователя, оставившего оценку;
   - rating — оценка книги.
   
- **Таблица reviews**
- Содержит данные о пользовательских обзорах:
   - review_id — идентификатор обзора;
   - book_id — идентификатор книги;
   - username — имя автора обзора;
   - text — текст обзора.

# Задания<a class="anchor" id="1"></a>

- [x] [Посчитать, сколько книг вышло после 1 января 2000 года](#books)

- [x] [Для каждой книги посчитать количество обзоров и среднюю оценку](#reviews)

- [x] [Определить издательство, которое выпустило наибольшее число книг толще 50 страниц](#over50)

- [x] [Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;](#over50raiting)

- [x] [Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.](#review48)



In [1]:
import pandas as pd
from sqlalchemy import create_engine
pd.options.display.max_colwidth = 1

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя 
'pwd': 'Sdf4$2;d-d30pp', # пароль 
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
'port': 6432, # порт подключения 
'db': 'data-analyst-final-project-db'} # название базы данных 
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config) 

In [3]:
# сохраняем коннектор 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [4]:
# чтобы выполнить SQL-запрос, используем Pandas 
"""" query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect() 
pd.io.sql.read_sql(sql=text(query), con = con) """


'" query = \'\'\'SELECT * FROM books LIMIT 5\'\'\'\ncon=engine.connect() \npd.io.sql.read_sql(sql=text(query), con = con) '

In [5]:
con=engine.connect()

In [6]:
def select(x):
  return pd.io.sql.read_sql(x, con = con)

# Исследование баз данных

In [7]:
books = '''
SELECT * 
FROM books
'''

In [8]:
authors = '''
SELECT * 
FROM authors
'''

In [9]:
publishers = '''
SELECT * 
FROM publishers
'''

In [10]:
ratings = '''
SELECT * 
FROM ratings
'''

In [11]:
reviews = '''
SELECT * 
FROM reviews
'''

In [12]:
select(books).head()

book_id  author_id                                                  title  \
0  1        546        'Salem's Lot                                            
1  2        465        1 000 Places to See Before You Die                      
2  3        407        13 Little Blue Envelopes (Little Blue Envelope  #1)     
3  4        82         1491: New Revelations of the Americas Before Columbus   
4  5        125        1776                                                    

   num_pages publication_date  publisher_id  
0  594        2005-11-01       93            
1  992        2003-05-22       336           
2  322        2010-12-21       135           
3  541        2006-10-10       309           
4  386        2006-07-04       268

In [13]:
db_list = ['books','authors','publishers','ratings','reviews']

In [14]:
for i in db_list:
    display(select(i).head())
    display(select(i).info())

book_id  author_id                                                  title  \
0  1        546        'Salem's Lot                                            
1  2        465        1 000 Places to See Before You Die                      
2  3        407        13 Little Blue Envelopes (Little Blue Envelope  #1)     
3  4        82         1491: New Revelations of the Americas Before Columbus   
4  5        125        1776                                                    

   num_pages publication_date  publisher_id  
0  594       2005-11-01        93            
1  992       2003-05-22        336           
2  322       2010-12-21        135           
3  541       2006-10-10        309           
4  386       2006-07-04        268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   book_id           1000 non-null   int64         
 1   author_id         1000 non-null   int64         
 2   title             1000 non-null   object        
 3   num_pages         1000 non-null   int64         
 4   publication_date  1000 non-null   datetime64[ns]
 5   publisher_id      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


None

author_id                          author
0  1          A.S. Byatt                    
1  2          Aesop/Laura Harris/Laura Gibbs
2  3          Agatha Christie               
3  4          Alan Brennert                 
4  5          Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

publisher_id                          publisher
0  1             Ace                              
1  2             Ace Book                         
2  3             Ace Books                        
3  4             Ace Hardcover                    
4  5             Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

rating_id  book_id       username  rating
0  1          1        ryanfranco     4     
1  2          1        grantpatricia  2     
2  3          1        brandtandrea   5     
3  4          2        lorichen       3     
4  5          2        mariokeller    2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

review_id  book_id       username  \
0  1          1        brandtandrea    
1  2          1        ryanfranco      
2  3          2        lorichen        
3  4          3        johnsonamanda   
4  5          3        scotttamara     

                                                                                                                               text  
0  Mention society tell send professor analysis. Over provide race technology continue these.                                        
1  Foot glass pretty audience hit themselves. Among admit investment argue security.                                                 
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3  Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.          
4  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

In [15]:
for i in db_list:
    display(select(i).isnull().sum())

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

author_id    0
author       0
dtype: int64

publisher_id    0
publisher       0
dtype: int64

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

review_id    0
book_id      0
username     0
text         0
dtype: int64

In [16]:
for i in db_list:
    display(select(i).duplicated().sum())

0

0

0

0

0

Вывод по выгрузке файлов:

- Таблица books
    - 1000 строк и 6 столбцов.

- Таблица authors
    - 636 строк и 2 столбца.

- Таблица publishers
    - 340 строк и 2 столбца.
    
- Таблица ratings
    - 6456 строк и 4 столбца.

- Таблица reviews
    - 2793 строк и 4 столбца.
    
- Пропуски и дубликаты в каждой таблице отсутстувуют.

# Задание 1<a class="anchor" id="books"></a>
Посчитать, сколько книг вышло после 1 января 2000 года.

[К оглавлению](#1)

In [17]:
count_books = '''
SELECT COUNT(book_id) 
FROM books
WHERE publication_date > '2000-01-01'
'''


In [18]:
select(count_books)

count
0  819

Вывод: После 1 января 2000 года вышло 819 книг.

# Задание 2<a class="anchor" id="reviews"></a>

Для каждой книги посчитать количество обзоров и среднюю оценку.

[К оглавлению](#1)

In [19]:
select(reviews).head(3)

review_id  book_id      username  \
0  1          1        brandtandrea   
1  2          1        ryanfranco     
2  3          2        lorichen       

                                                                                                                               text  
0  Mention society tell send professor analysis. Over provide race technology continue these.                                        
1  Foot glass pretty audience hit themselves. Among admit investment argue security.                                                 
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.

In [20]:
select(ratings).head(3)

rating_id  book_id       username  rating
0  1          1        ryanfranco     4     
1  2          1        grantpatricia  2     
2  3          1        brandtandrea   5

In [21]:
count_reviews = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT r.text) AS count_reviews,
       RoUND(AVG(ra.rating),2) AS average_rating
FROM books AS b
LEFT JOIN reviews AS r ON r.book_id = b.book_id
LEFT JOIN ratings AS ra ON ra.book_id = b.book_id
GROUP BY b.book_id,
         b.title
ORDER BY count_reviews DESC '''

In [22]:
select(count_reviews)

book_id                                                        title  \
0    948      Twilight (Twilight  #1)                                       
1    963      Water for Elephants                                           
2    734      The Glass Castle                                              
3    302      Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)   
4    695      The Curious Incident of the Dog in the Night-Time             
..   ...                                                    ...             
995  83       Anne Rice's The Vampire Lestat: A Graphic Novel               
996  808      The Natural Way to Draw                                       
997  672      The Cat in the Hat and Other Dr. Seuss Favorites              
998  221      Essential Tales and Poems                                     
999  191      Disney's Beauty and the Beast (A Little Golden Book)          

     count_reviews  average_rating  
0    7              3.66            
1    6              3.98            
2    6              4.21            
3    6              4.41            
4    6              4.08            
..  ..               ...            
995  0              3.67            
996  0              3.00            
997  0              5.00            
998  0              4.00            
999  0              4.00            

[1000 rows x 4 columns]

Вывод: 
Была выведена таблица с кол-во обзоров и средним рейтингом для каждой книги. 

Самая обозреваемая книга Twilight (Twilight #1)	с рейтингом 3.66.

# Задание 3<a class="anchor" id="over50"></a>

Определить издательство, которое выпустило наибольшее число книг толще 50 страниц.

[К оглавлению](#1)

In [23]:
select(publishers)

publisher_id                          publisher
0    1             Ace                              
1    2             Ace Book                         
2    3             Ace Books                        
3    4             Ace Hardcover                    
4    5             Addison Wesley Publishing Company
..  ..                                           ...
335  336           Workman Publishing Company       
336  337           Wyatt Book                       
337  338           Yale University Press            
338  339           Yearling                         
339  340           Yearling Books                   

[340 rows x 2 columns]

In [24]:
publisher_over_50 = '''
SELECT p.publisher,
       COUNT(b.book_id) AS count_books
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY count_books DESC
LIMIT 3
'''

In [25]:
select(publisher_over_50)

publisher  count_books
0  Penguin Books             42         
1  Vintage                   31         
2  Grand Central Publishing  25

Вывод:
 
 Издательство Penguin Books выпустило больше всего книг(42) толще 50 страниц.

# Задание 4<a class="anchor" id="over50raiting"></a>

Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

[К оглавлению](#1)


In [26]:
select(authors).head(1)

author_id      author
0  1          A.S. Byatt

In [27]:
select(ratings).head(1)

rating_id  book_id    username  rating
0  1          1        ryanfranco  4

In [28]:
select(books).head(1)

book_id  author_id         title  num_pages publication_date  publisher_id
0  1        546        'Salem's Lot  594        2005-11-01       93

In [29]:
author_over_50 = '''
SELECT a.author,
       ROUND(AVG(r.rating), 3) AS average_rating
FROM books AS b
INNER JOIN authors AS a ON a.author_id = b.author_id
LEFT JOIN ratings AS r ON r.book_id = b.book_id
GROUP BY a.author
HAVING COUNT(rating) > 50
ORDER BY average_rating DESC
LIMIT 5
'''

In [30]:
select(author_over_50)

author  average_rating
0  J.K. Rowling/Mary GrandPré         4.288         
1  Agatha Christie                    4.283         
2  Markus Zusak/Cao Xuân Việt Khương  4.264         
3  J.R.R. Tolkien                     4.241         
4  Roald Dahl/Quentin Blake           4.210

Вывод: Автор с самой высокой средней оценкой книг, где оценок больше 50 - это J.K. Rowling/Mary GrandPré.

# Задание 5<a class="anchor" id="review48"></a>

Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

[К оглавлению](#1)


In [31]:
select(reviews).head(1)

review_id  book_id      username  \
0  1          1        brandtandrea   

                                                                                         text  
0  Mention society tell send professor analysis. Over provide race technology continue these.

In [32]:
select(ratings).head(1)

rating_id  book_id    username  rating
0  1          1        ryanfranco  4

In [33]:
users ='''
SELECT AVG(count_reviews) AS average_count_reviews
FROM
  (SELECT r.username,
          COUNT(text) AS count_reviews
   FROM reviews AS r
   JOIN
     (SELECT username,
             COUNT(rating_id) AS count_rating
      FROM ratings
      GROUP BY username) AS cr ON cr.username = r.username
   WHERE count_rating > 48
   GROUP BY r.username) AS avg_count
'''

In [34]:
select(users)

average_count_reviews
0  24.0

Вывод: В среднем пользователи с кол-во оценок больше 48 оставляют 24 обзора. То есть на каждую 2 оценку пишется обзор.

# Общий вывод по заданиям

[К оглавлению](#1)

1. После 1 января 2000 года вышло 819 книг.


2. Была выведена таблица(count_reviews) с кол-во обзоров и средним рейтингом для каждой книги.

    2.1 Самая обозреваемая книга Twilight (Twilight #1) с рейтингом 3.66.
    

3. Издательство Penguin Books выпустило больше всего книг(42) толще 50 страниц.


4. Автор с самой высокой средней оценкой книг, где оценок больше 50 - это J.K. Rowling/Mary GrandPré.


5. В среднем пользователи с кол-во оценок больше 48 оставляют 24 обзора. То есть на каждую 2 оценку пишется обзор.